lib/music_theory/note.rb

```diff
@@ -37,10 +38,41 @@ def samples
         phase += sine_wave_cycle
         sample
       end
+      samples = envelope!(samples) if envelope
       samples = distort!(samples) if distort
       samples
     end
 
+    def envelope!(samples)
+      # some function that filters amplitude over the duration of a given note - (to be expanded on)
+      # attack should go from 0 amplitude to max over the course of the first 1/4 of note duration
+      # sustain for the second quarter of the note duration
+      # decay should take the rest of the duration (3/4 and 4/4) of the note to return to 0
+      max_amplitude = samples.map {|s| s.abs }.max
+      slice_size = samples.length / 4
+      attack_slice = samples[0..slice_size]
+      sustain_slice = samples[slice_size..(slice_size * 2)]
+      decay_slice = samples[(slice_size * 2)..(slice_size * 4)]
+
+      amplitude = 0
+      attack_incrementor = max_amplitude / slice_size
+      attack = attack_slice.each do |sample|
+        sample = amplitude
+        amplitude += attack_incrementor
+      end
+
+      sustain = sustain_slice.map {|sample| sample = max_amplitude}
+
+      amplitude = max_amplitude
+      decay_incrementor = max_amplitude / (slice_size * 2)
+      decay = decay_slice.each do |sample|
+        sample = amplitude
+        amplitude -= decay_incrementor
+      end
+
+      samples = [attack, sustain, decay].flatten
+    end
```

Comment:
@bion Here's my terribly imperative idea of how an envelope should work - thinking I may be missing some key concepts here but all I've gleaned from my research is that physicists aren't good at communicating... please advise :)